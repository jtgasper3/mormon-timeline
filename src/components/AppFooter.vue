<template>
  <v-footer app>
    <v-card elevation="0" rounded="0" width="100%" class="bg-grey text-center">
      <v-card-text>
        <v-btn class="mx-4" icon="mdi-home" variant="plain"></v-btn>
        <v-btn class="mx-4" icon="mdi-email" variant="plain"></v-btn>
        <v-btn class="mx-4" icon="mdi-calendar" variant="plain"></v-btn>
      </v-card-text>

      <v-divider></v-divider>

      <v-card-text class="text-white">
        {{ new Date().getFullYear() }} â€” <strong>John Gasper - v{{ version }}</strong>
      </v-card-text>
    </v-card>
  </v-footer>
</template>

<script setup>
  // eslint-disable-next-line no-undef
  const version = __APP_VERSION__
</script>