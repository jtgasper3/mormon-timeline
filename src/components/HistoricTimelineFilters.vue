<template>
  <v-card>
    <v-toolbar
      dark
      color="primary"
    >
      <v-btn
        icon
        dark
        @click="close()"
      >
        <v-icon>mdi-close</v-icon>
      </v-btn>
      <v-toolbar-title>Filters</v-toolbar-title>
      <v-spacer></v-spacer>
      <v-toolbar-items>
        <v-btn
          dark
          text
          @click="close()"
        >
          Save
        </v-btn>
      </v-toolbar-items>
    </v-toolbar>
  
    <v-expansion-panels>
      <v-expansion-panel>
        <v-expansion-panel-title> Denominations </v-expansion-panel-title>
        <v-expansion-panel-text>
          <v-row no-gutters>
            <v-col v-for="item in denominations" :key="item.value" cols="4">
              <v-switch
                v-model="filters.denominations"
                :label="item.label"
                :value="item.value"
                hide-details
                density="compact"
              ></v-switch>
            </v-col>
          </v-row>
        </v-expansion-panel-text>
      </v-expansion-panel>
      <v-expansion-panel>
        <v-expansion-panel-title> Categories </v-expansion-panel-title>
        <v-expansion-panel-text>
          <v-row no-gutters>
            <v-col v-for="item in categories" :key="item.value" cols="4">
              <v-switch
                v-model="filters.categories"
                :color="item.color"
                :label="item.label"
                :value="item.value"
                hide-details
                density="compact"
              ></v-switch>
            </v-col>
          </v-row>
        </v-expansion-panel-text>
      </v-expansion-panel>
      <v-expansion-panel>
        <v-expansion-panel-title> Dates </v-expansion-panel-title>
        <v-expansion-panel-text>
          <v-row no-gutters>
            <v-col cols="12">
              Time Range:
              <v-range-slider
                v-model="filters.dateRange"
                :min="filters.dateRangeMin"
                :max="filters.dateRangeMax"
                step="10"
                thumb-label="always"
                strict
              ></v-range-slider>
            </v-col>
          </v-row>
        </v-expansion-panel-text>
      </v-expansion-panel>
    </v-expansion-panels>
  </v-card>
</template>

<script setup>
import { ref } from "vue";

const props = defineProps({
  filters: { 
    type: Object,
    required: true
  },
})

const filters = ref(props.filters)
const emit = defineEmits(['close', 'change'])

const denominations = [
  {
    label: 'Common',
    value: 'common',
  },
  {
    label: 'Latter-day Saints',
    value: 'lds',
  },
  {
    label: 'Community of Christ',
    value: 'rlds',
  },
]

const categories = [
  {
    label: "Book Of Mormon",
    value: "bom",
    color: "cyan",
  },
  {
    label: "People",
    value: "people",
    color: "pink",
  },
]; // amber, orange

function close() {
  emit("close");
}

</script>
