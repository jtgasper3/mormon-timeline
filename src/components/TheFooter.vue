<template>
  <v-footer
    app
    absolute
    class="bg-grey-lighten-1"
  >
    <v-row
      justify="center"
      no-gutters
    >
      <v-btn
        v-for="link in links"
        :key="link"
        color="white"
        variant="text"
        class="mx-2"
        rounded="xl"
        :to="link.link"
      >
        {{ link.label }}
      </v-btn>
      <v-col
        class="text-center text-white mt-4"
        cols="12"
      >
        Managed by John Gasper - v{{ version }} - Built on {{ buildDate }}
      </v-col>
    </v-row>
  </v-footer>
</template>

<script setup>
  // eslint-disable-next-line no-undef
  const version = __APP_VERSION__
  // eslint-disable-next-line no-undef
  const buildDate = __APP_BUILD_DATE__

  const links = [
    { label: 'Home', link: '/' },
    { label: 'About', link: 'about' },
    { label: 'Team', link: 'team' },
    { label: 'Contact Us', link: 'contact' }
  ]
</script>