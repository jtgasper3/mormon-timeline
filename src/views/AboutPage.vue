<template>
  <v-row>
    <v-col>

    </v-col>
  </v-row>
</template>

<script setup>
// import { reactive } from "vue";

// const state = reactive({

// });
</script>
