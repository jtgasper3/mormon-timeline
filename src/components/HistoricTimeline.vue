<template>
  <v-row>
    <v-col>
      <v-text-field
        v-model="state.searchString"
        variant="contained"
        hint="Search the timeline for specific text. (Other filters are available from the Filter menu.)"
        prepend-inner-icon="mdi-magnify"
        label="Search"
      />
    </v-col>
  </v-row>
  <v-row>
    <v-col>
      <v-timeline>
        <v-timeline-item
          v-for="event in filteredEvents"
          :key="event.id"
          fill-dot
          :dot-color="getCategoryColor(event.category)"
          size="small"
        >
          <template #opposite>
            <div
              :class="`pt-1 headline font-weight-bold text-${getCategoryColor(
                event.category
              )}`"
            >
              {{ formatDate(event.date) }}
            </div>
          </template>

          <div>
            <h2
              v-if="event.title"
              :class="`mt-n1 headline font-weight-light mb-4 text-${getCategoryColor(
                event.category
              )} justify-end`"
            >
              {{ event.title }}
            </h2>
            <div v-if="event.subtitle">{{ event.subtitle }}</div>
            <div v-if="event.description">
              {{ event.description }}
            </div>
          </div>
        </v-timeline-item>
      </v-timeline>
    </v-col>
  </v-row>

  <v-dialog
    v-model="state.showFilters"
    fullscreen
    :scrim="false"
    transition="dialog-bottom-transition"
  >
    <template #activator="{ props }">
      <v-card
        class="px-1"
        flat
        fixed
        top="104"
        right="-35"
        width="90"
        color="rgba(0, 0, 0, 0.1)"
        theme="dark"
      >
        <v-tooltip anchor="start">
          <template #activator="{ props: tipProps }">
            <v-btn
              aria-label="Filter"
              color="primary"
              icon=""
              theme="dark"
              variant="plain"
              v-bind="props"
            >
              <v-icon color="primary" dark v-bind="tipProps">
                mdi-filter-menu
              </v-icon>
            </v-btn>
          </template>
          <span>Filter</span>
        </v-tooltip>
      </v-card>
    </template>
    <historic-timeline-filters
      :filters="state.filters"
      @close="state.showFilters = false"
    ></historic-timeline-filters>
  </v-dialog>
</template>

<script setup>
import { computed, onMounted, reactive, watch } from "vue";
import { format } from "date-fns";
import HistoricTimelineFilters from "@/components/HistoricTimelineFilters.vue";
import { getEvents } from "@/services";

const categories = [
  {
    label: "Book Of Mormon",
    value: "bom",
    color: "cyan",
  },
  {
    label: "People",
    value: "people",
    color: "pink",
  },
]; // amber, orange

const state = reactive({
  events: [],
  filters: {
    categories: [],
    denominations: [],
    dateRange: [],
    dateRangeMin: 0,
    dateRangeMax: 0,
  },
  searchString: "",
  showFilters: false,
});

onMounted(() => {
  state.events = getEvents() ?? [];
});

const filteredEvents = computed(() => {
  let searchStringPattern;

  try {
    searchStringPattern = new RegExp(state.searchString, "i");
  } catch (e) {
    // searchStringPattern will be falsy
  }

  const initialFiltered = state.events
    .filter((e) => state.filters.categories.includes(e.category))
    .filter((e) => state.filters.denominations.includes(e.denomination))
    .filter(
      (e) =>
        e.date > new Date(state.filters.dateRange[0], 0) &&
        e.date < new Date(state.filters.dateRange[1], 11)
    );
  if (!searchStringPattern) {
    return initialFiltered;
  }

  return initialFiltered.filter(
    (e) =>
      searchStringPattern.test(e.title) ||
      searchStringPattern.test(e.subtitle) ||
      searchStringPattern.test(e.description)
  );
});

function getCategoryColor(category) {
  const categoryObj = categories.find((i) => i.value === category);
  return !categoryObj ? "black" : categoryObj.color;
}

watch(
  () => state.events,
  () => {
    state.filters.dateRangeMin =
      Math.floor(
        Math.min.apply(
          Math,
          state.events.map(function (o) {
            return new Date(o.date).getUTCFullYear();
          })
        ) / 10
      ) * 10;
    state.filters.dateRangeMax =
      Math.ceil(
        Math.max.apply(
          Math,
          state.events.map(function (o) {
            return new Date(o.date).getUTCFullYear();
          })
        ) / 10
      ) * 10;
    if (state.filters.dateRange.length === 0) {
      state.filters.dateRange = [
        state.filters.dateRangeMin,
        state.filters.dateRangeMax,
      ];
    }

    if (state.filters.categories.length === 0) {
      state.filters.categories = state.events
        .map((e) => e.category)
        .filter((category, index, arr) => arr.indexOf(category) == index);
    }

    if (state.filters.denominations.length === 0) {
      state.filters.denominations = state.events
        .map((e) => e.denomination)
        .filter(
          (denominations, index, arr) => arr.indexOf(denominations) == index
        );
    }
  }
);

function formatDate(date) {
  return format(date, "PPP");
}
</script>
