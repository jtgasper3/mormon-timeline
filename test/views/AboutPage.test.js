import { nextTick } from "vue";
import { describe, it, expect } from "vitest";
import { mount } from "@vue/test-utils";
import { createVuetify } from "vuetify";
import * as components from "vuetify/components";
import * as directives from "vuetify/directives";
import AboutPage from "@/views/AboutPage.vue";

describe("AboutPage", () => {
  const vuetify = createVuetify({ components, directives });

  it("mount component", async () => {
    expect(AboutPage).toBeTruthy();

    const wrapper = mount(AboutPage, {
      global: {
        plugins: [vuetify],
      },
    });

    await nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
