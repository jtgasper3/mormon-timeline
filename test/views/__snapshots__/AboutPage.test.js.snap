// Vitest Snapshot v1

exports[`AboutPage > mount component 1`] = `
"<div class=\\"v-row\\">
  <div class=\\"v-col\\"></div>
</div>"
`;
